crates/containerd-shim-wasm/src/sandbox/cgroups.rs

```diff
@@ -0,0 +1,674 @@
+            }
+        }
+
+        if let Some(blkio) = res.block_io() {
+            if let Some(weight) = blkio.weight() {
+                ensure_write_file(
+                    self.get_file("io.weight"),
+                    &format!("{}", (weight - 10) * 9999 / 990),
+                )?;
+            }
+
+            if let Some(throttle_write_bps_device) = blkio.throttle_read_bps_device() {
+                for device in throttle_write_bps_device {
+                    let path = self.get_file(&format!("io.max"));
+                    ensure_write_file(
+                        path,
+                        &format!(
+                            "{}:{} rbps={}",
+                            device.major(),
+                            device.minor(),
+                            device.rate()
+                        ),
+                    )?;
+                }
+            }
+
+            if let Some(throttle_write_bps_device) = blkio.throttle_write_bps_device() {
+                for device in throttle_write_bps_device {
+                    let path = self.get_file(&format!("io.max"));
+                    ensure_write_file(
+                        path,
+                        &format!(
+                            "{}:{} wbps={}",
+                            device.major(),
+                            device.minor(),
+                            device.rate()
+                        ),
+                    )?;
+                }
+            }
+
+            if let Some(throttle_write_bps_device) = blkio.throttle_read_iops_device() {
+                for device in throttle_write_bps_device {
+                    let path = self.get_file(&format!("io.max"));
+                    ensure_write_file(
+                        path,
+                        &format!(
+                            "{}:{} riops={}",
+                            device.major(),
+                            device.minor(),
+                            device.rate()
+                        ),
+                    )?;
+                }
+            }
+
+            if let Some(throttle_write_bps_device) = blkio.throttle_write_iops_device() {
+                for device in throttle_write_bps_device {
+                    let path = self.get_file(&format!("io.max"));
+                    ensure_write_file(
+                        path,
+                        &format!(
+                            "{}:{} wiops={}",
+                            device.major(),
+                            device.minor(),
+                            device.rate()
+                        ),
+                    )?;
+                }
+            }
+        }
+
+        if let Some(unified) = res.unified() {
+            for (k, v) in unified {
+                ensure_write_file(self.get_file(k), v)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub fn new(name: String) -> Result<Box<dyn Cgroup>> {
```

Comment:
If would be great to see more unit tests for `new` function, considering its logical to determine v1 and v2 cgroups is quiet obscuring. 

For fast path, we probably need 
- one test case for detecting hybrid mode, 
- one test case with v2 mounted on v1 path, 
- one test case with v1 mounted on v2 path,
- one test case where v1 v2 both not found

Reply:
Not asking these unit tests in this PR though. I'd like to create backlog and have another PR to increase the test coverage.

Reply:
Indeed this function is tricky... and even trickier to test in CI because we can't really easily control different cgroup modes.